lib/src/client_offline_mixin.dart

```diff
@@ -0,0 +1,310 @@
+  //   final record = await recordRef.get(db);
+  //   int change = 0;
+  //   if (record == null) {
+  //     final encoded = encode(map);
+  //     change = encoded.length;
+  //   } else {
+  //     change = calculateChange(record, map);
+  //   }
+
+  //   await updateCacheSize(db, change);
+  //   await recordRef.put(db, map);
+  //   await updateAccessedAt(db, store.name, key);
+  // }
+
+  // Future<void> upsertCacheOld(
+  //     DatabaseClient db,
+  //     StoreRef<String, Map<String, Object?>> store,
+  //     String id,
+  //     Map<String, dynamic> map) async {
+  //   final record = await store.findFirst(db,
+  //       finder: Finder(filter: Filter.equals('\$id', id)));
+
+  //   if (record == null) {
+  //     final encoded = encode(map);
+  //     final change = encoded.length;
+  //     await updateCacheSize(db, change);
+  //     final key = await store.add(db, map);
+  //     await updateAccessedAt(db, store.name, key);
+  //     return;
+  //   }
+
+  //   final updated = await record.ref.update(db, map);
+  //   if (updated != null) {
+  //     final change = calculateChange(record.value, map);
+  //     await updateCacheSize(db, change);
+  //   }
+  //   await updateAccessedAt(db, store.name, record.key);
+  // }
+
+  Future<void> upsertCache(DatabaseClient db,
+      StoreRef<String, Map<String, Object?>> store, Map<String, dynamic> map,
+      {String? key, String? id}) async {
+    if (key == null && id == null) {
+      throw AppwriteException(
+          'key and id cannot be null', 0, 'general_cache_error');
+    }
+
+    if (key != null) {
+      final recordRef = store.record(key);
+      final record = await recordRef.get(db);
+      int change = 0;
+      if (record == null) {
+        final encoded = encode(map);
+        change = encoded.length;
+      } else {
+        change = calculateChange(record, map);
+      }
+
+      await updateCacheSize(db, change);
+      await recordRef.put(db, map);
+      await updateAccessedAt(db, store.name, key);
+      return;
+    }
+
+    final record = await store.findFirst(db,
+        finder: Finder(filter: Filter.equals('\$id', id)));
+
+    if (record == null) {
+      final encoded = encode(map);
+      final change = encoded.length;
+      await updateCacheSize(db, change);
+      final key = await store.add(db, map);
+      await updateAccessedAt(db, store.name, key);
+      return;
+    }
+
+    final updated = await record.ref.update(db, map);
+    if (updated != null) {
+      final change = calculateChange(record.value, map);
+      await updateCacheSize(db, change);
+    }
+    await updateAccessedAt(db, store.name, record.key);
+  }
+
+  Future<void> deleteCache(
+      DatabaseClient db, StoreRef<String, Map<String, Object?>> store,
+      {String? key, String? id}) async {
+    if (key == null && id == null) {
+      throw AppwriteException(
+          'key and id cannot be null', 0, 'general_cache_error');
+    }
```

Comment:
IF this cannot be null, why don't we  make it required attribute?

Reply:
Because only 1 needs to be required rather than both.